{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5cd48355-5ca5-4031-8f3e-93e083f7bf23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_11135.fasta\n",
      "Number in Dros seq: 80  Or at a 5th rate: 16.0\n",
      "Number in mammals: 83.93486590038314\n",
      "CpGs per 1000 in Drosophila seq: 36.98566805362922\n",
      "CpGs per 1000 in mammals seq: 59.897414024093486\n",
      "gene_12690.fasta\n",
      "Number in Dros seq: 37  Or at a 5th rate: 7.4\n",
      "Number in mammals: 64.95019157088123\n",
      "CpGs per 1000 in Drosophila seq: 30.98827470686767\n",
      "CpGs per 1000 in mammals seq: 59.78339452032572\n",
      "gene_15823.fasta\n",
      "Number in Dros seq: 74  Or at a 5th rate: 14.8\n",
      "Number in mammals: 51.030651340996165\n",
      "CpGs per 1000 in Drosophila seq: 71.4975845410628\n",
      "CpGs per 1000 in mammals seq: 59.902403921833184\n",
      "gene_17683.fasta\n",
      "Number in Dros seq: 59  Or at a 5th rate: 11.8\n",
      "Number in mammals: 90.2375478927203\n",
      "CpGs per 1000 in Drosophila seq: 29.265873015873016\n",
      "CpGs per 1000 in mammals seq: 60.0451766529505\n",
      "gene_19179.fasta\n",
      "Number in Dros seq: 47  Or at a 5th rate: 9.4\n",
      "Number in mammals: 59.55555555555556\n",
      "CpGs per 1000 in Drosophila seq: 34.81481481481481\n",
      "CpGs per 1000 in mammals seq: 60.70760446324308\n",
      "gene_5927.fasta\n",
      "Number in Dros seq: 57  Or at a 5th rate: 11.4\n",
      "Number in mammals: 51.02298850574713\n",
      "CpGs per 1000 in Drosophila seq: 30.158730158730158\n",
      "CpGs per 1000 in mammals seq: 56.17778452738018\n",
      "gene_5978.fasta\n",
      "Number in Dros seq: 27  Or at a 5th rate: 5.4\n",
      "Number in mammals: 51.559386973180075\n",
      "CpGs per 1000 in Drosophila seq: 27.35562310030395\n",
      "CpGs per 1000 in mammals seq: 55.60031566204329\n",
      "gene_7441.fasta\n",
      "Number in Dros seq: 48  Or at a 5th rate: 9.6\n",
      "Number in mammals: 51.60919540229885\n",
      "CpGs per 1000 in Drosophila seq: 32.454361054766736\n",
      "CpGs per 1000 in mammals seq: 55.2662998658346\n",
      "gene_8308.fasta\n",
      "Number in Dros seq: 18  Or at a 5th rate: 3.6\n",
      "Number in mammals: 31.436781609195403\n",
      "CpGs per 1000 in Drosophila seq: 31.088082901554404\n",
      "CpGs per 1000 in mammals seq: 55.20456976767656\n"
     ]
    }
   ],
   "source": [
    "### This is the code for the drosophila control test\n",
    "\n",
    "#important! you will need to get the homologue gene sequence for your genes of choice from NCBI\n",
    "#Also the drosophila sequence text file must be named the same as the respective mammalian gene file; gene_XXX.fasta\n",
    "#The files will need to be in a folder called: 'dros_gene'\n",
    "\n",
    "#When I ran this test, I would change the files that are in the dros_gene folder when I wanted to look at higher or lower CpG genes\n",
    "#It is possible to keep all the genes and run them together, before seperating them out afterwards. This is up to you.\n",
    "\n",
    "\n",
    "#load packages\n",
    "import os\n",
    "\n",
    "#This fuction pulls the sequence data (sometime multiple sequences) from the inputted text files\n",
    "def seqpull(file):\n",
    "    with open(file, 'r') as f: \n",
    "        lines=f.read()\n",
    "        lines=lines.split('\\n')\n",
    "        sequence=''\n",
    "        sequence_dic={}\n",
    "        first=0\n",
    "        for line in lines:\n",
    "            if first==0:\n",
    "                species=line\n",
    "                first=1\n",
    "            elif len(line)==0:\n",
    "                continue\n",
    "            elif line[0]=='>':\n",
    "                sequence_dic[species]=sequence\n",
    "                species=line\n",
    "                sequence=''\n",
    "            else:\n",
    "                sequence=sequence+line.upper()        \n",
    "        sequence_dic[species]=sequence\n",
    "        return(sequence)\n",
    "    \n",
    "#This counts the number of CpGs in the inputted sequence\n",
    "def dyad_counter(data):\n",
    "        i = 0\n",
    "        CpG = 0\n",
    "        dyad_locations = []\n",
    "        dyad_seq=[]\n",
    "        #A while loop that goes up in codons and checks between the codons for CpGs, it then counts these.\n",
    "        while i< len(data):\n",
    "            #counts up the index each loop\n",
    "            i=i+3\n",
    "            #checks if the index is above the size of the sequence after the addition. \n",
    "            if i>=len(data):\n",
    "                break\n",
    "        #checks if both nessecary bases are C and G, if so it adds 1 to the count. \n",
    "            if (data[i]=='G' and data[i-1]=='C'):#change this if you are not interested in dyads. \n",
    "                CpG+=1\n",
    "                \n",
    "        return(CpG)\n",
    "    \n",
    "#This automates the generation of regular gene names for later use\n",
    "with open('mouse_to_human.txt','r') as f:\n",
    "    lines=f.read()\n",
    "    lines=lines.split('\\n')\n",
    "    my_dic={}\n",
    "    for line in lines[1:]:\n",
    "        line=line.split('\\t')\n",
    "        if len(line)<4 or line[2]=='' or line[3]=='':\n",
    "            continue\n",
    "        elif line[3] not in ens_list:\n",
    "            continue\n",
    "        else:\n",
    "            my_dic[line[3]]=line[2]\n",
    "#this pulls the general information for each gene sequence\n",
    "#It is not very efficient, as it does all genes, when really we are only interested in a few.\n",
    "with open('General_summary.csv', 'r') as data:\n",
    "    values_dic={}\n",
    "    length_dic={}\n",
    "    lines = data.read()\n",
    "    lines = lines.split('\\n')\n",
    "    file_dic={}\n",
    "\n",
    "    for line in lines[1:]:\n",
    "        line=line.split(',')\n",
    "        if len(line)==1:\n",
    "            continue\n",
    "        else:\n",
    "            values_dic[line[0]]=float(line[3])\n",
    "            length_dic[line[0]]=float(line[4])\n",
    "            file_dic[line[0]]=line[1]\n",
    "            \n",
    "            \n",
    "#This creates the comparison file. I renamed the file here each time i edited the selection of genes\n",
    "with open('dros_comparison_high.txt','w') as res:\n",
    "    res.write('file \\tspec \\tval \\n')\n",
    "    for file in os.listdir('dros_genes/'):\n",
    "        if file in file_dic.keys() and file_dic[file] in my_dic.keys():\n",
    "            print(file)\n",
    "            #This line pulls the sequence from the dros file\n",
    "            my_seq = seqpull('dros_genes/'+file)\n",
    "            #This line gets the number of CpGs in the sequence\n",
    "            my_cpg=dyad_counter(my_seq)\n",
    "            #These lines print out some informtion for later use, as well as write the data to a result file.\n",
    "            print('Number in Dros seq: '+str(my_cpg)+'  Or at a 5th rate: '+str(my_cpg/5))\n",
    "            print('Number in mammals: '+str(values_dic[file]))\n",
    "            cpg1000dros = (my_cpg/len(my_seq))*1000\n",
    "            print('CpGs per 1000 in Drosophila seq: '+str(cpg1000dros))\n",
    "            cpg1000mal = (values_dic[file]/length_dic[file])*1000\n",
    "            print('CpGs per 1000 in mammals seq: '+str(cpg1000mal))\n",
    "            res.write(my_dic[file_dic[file]]+'\\t'+'dros'+'\\t'+str(cpg1000dros)+'\\n')\n",
    "            res.write(my_dic[file_dic[file]]+'\\t'+'mal'+'\\t'+str(cpg1000mal)+'\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
