{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a1a583-3736-42ac-a3ff-a2610a2553a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "### Analyse all data in 'Test data' ###\n",
    "###################################\n",
    "\n",
    "# This script will analyse all sequences in the 'test data' folder.\n",
    "# Currently it is only set up for this specific data set, meaning a FASTA file named in the way the example data is\n",
    "# final results for individual genes are outputted into the test_results folder, a few general result documents are also generated, the are outputted to the source folder.\n",
    "# The code could be adapted to look for .txt files or other files with some effort. \n",
    "\n",
    "### Instructions!\n",
    "\n",
    "# This is very simple, put all test data into the test_data folder, then hit ctrl+enter\n",
    "\n",
    "# The only thing that is worth mentioning is that to match the ENSG gene code to the gene number there is a gene_correspondents.txt file,\n",
    "# If you add new sequences that do not have a gene ID then you will likely get an error. I can edit the code so that it gives an 'NA' rather then an error if needed. \n",
    "\n",
    "#You can enter an extension name for the final results files if you like, otherwise it can be left blank\n",
    "extension='NNCGNN_'\n",
    "\n",
    "# This code is split into 4 sections, the first three are the same in all three codes. \n",
    "\n",
    "\n",
    "# First is the function used for identifying dyads\n",
    "# I recommend against changing this unless you want the general information document to present something different\n",
    "# If you do add or change data, you will need to extract it in the second section\n",
    "\n",
    "#Needed for generating the csv\n",
    "import csv\n",
    "#os is needed to iterate over the folder so you can do all data points there in \n",
    "import os\n",
    "#A counting package needed further on. \n",
    "from collections import Counter\n",
    "\n",
    "#Many genes have a double ATG at the start (nessecary in the previous study, but unwanted here)\n",
    "#This function checks if the gene has the double ATG, and records it. This allows for correction later on.\n",
    "def atg_checker(file):\n",
    "    with open('test_data/'+file) as f2:\n",
    "        lines=f2.read()\n",
    "        lines=lines.split('\\n')\n",
    "        count2=0\n",
    "        for line in lines:\n",
    "            if len(line)==0 or line[0]=='>':\n",
    "                continue\n",
    "            elif line[0:6]=='ATGATG' or line[0:6] =='ATG---':\n",
    "                count2+=1\n",
    "            else:\n",
    "                continue\n",
    "        if count2 ==261:\n",
    "            return(3)\n",
    "        else:\n",
    "            return(0)\n",
    "\n",
    "\n",
    "#Function for counting Dyads, more functionality can easily be added if needed. \n",
    "def dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg):\n",
    "        #Create/reset objects\n",
    "        i = 0\n",
    "        CpG = 0\n",
    "        dyad_locations = []\n",
    "        dyad_seq=[]\n",
    "        line_num=''\n",
    "        gene_ID=''\n",
    "        missing=0\n",
    "        #A while loop that goes up in codons and checks between the codons for CpGs, it then counts these.\n",
    "        while i< len(data):\n",
    "            #counts up the index each loop\n",
    "            i=i+3\n",
    "            #checks if the index is above the size of the sequence after the addition. \n",
    "            if i>=len(data):\n",
    "                break\n",
    "        #checks if both nessecary bases are C and G, if so it adds 1 to the count. \n",
    "            if (data[i]=='G' and data[i-1]=='C'): #or (data[i]=='G' and data[i-1]=='G' and data[i-2]=='C'): #or (data[i-1]=='G' and data[i-2]=='C') #edit depending on what you want to analyse.\n",
    "                CpG+=1\n",
    "                codons = data[i-3:i+3]\n",
    "                \n",
    "                #This section updates the 3 dictionaries with the data we need for analysing the codons. \n",
    "                #If the dyad position is already in the main dicitonary then it needs to update the relevant dictionaries\n",
    "                if i-atg in dyad_dic:\n",
    "                    #These lines add the species name to the species name dictionary, this data is mainly visual for the end\n",
    "                    temp_name_list = dyad_species_name[i-atg]\n",
    "                    temp_name_list.append(species_name)\n",
    "                    temp_name_dic = {i-atg:temp_name_list}\n",
    "                    dyad_species_name.update(temp_name_dic)\n",
    "                    #The next dictionary check is to see if the location's sequence is already present, if codons are conserved then is will happen a lot, it adds to the frequency then moves to the next codon\n",
    "                    if codons in dyad_dic[i-atg]:\n",
    "                        dyad_freq[i-atg]+=1\n",
    "                        second_temp=dyad_ratio_dic[i-atg]                                \n",
    "                        second_temp.append(codons)\n",
    "                        #print(second_temp)\n",
    "                        temp_dic = {i-atg:second_temp}\n",
    "                        dyad_ratio_dic.update(temp_dic)\n",
    "                        continue\n",
    "                    #If the codon is not present, then a new codon is added to entry at the given position\n",
    "                    else:\n",
    "                        temp_list=[]\n",
    "                        second_temp=[]\n",
    "                        for num in dyad_dic.items():\n",
    "                            if num[0]==i-atg:\n",
    "                                dyad_freq[i-atg]+=1\n",
    "                                temp_list=dyad_dic[i-atg]\n",
    "                                temp_list.append(codons)\n",
    "                                temp_dic = {i-atg:temp_list}\n",
    "                                dyad_dic.update(temp_dic)\n",
    "                                second_temp=dyad_ratio_dic[i-atg]                                \n",
    "                                second_temp.append(codons)\n",
    "                                #print(second_temp)\n",
    "                                temp_dic = {i-atg:second_temp}\n",
    "                                dyad_ratio_dic.update(temp_dic)\n",
    "                                \n",
    "                #If the entry is new, then the first set of data for the three dictionaries is entered.                \n",
    "                else:\n",
    "                    dyad_dic[i-atg]=[codons]\n",
    "                    dyad_freq[i-atg]=1\n",
    "                    dyad_species_name[i-atg] = [species_name]\n",
    "                    dyad_ratio_dic[i-atg]=[codons]\n",
    "                    \n",
    "        #This code turns the file_ID into a number, it then uses that number to get teh corresponding line in gene_correspondents, which then gives it the gene code which is recorded. \n",
    "        if file_ID !='':\n",
    "            line_num = file_ID.strip('.fasta')\n",
    "            line_num = int(line_num.strip('gene_'))\n",
    "\n",
    "            target_line = content[line_num-1]\n",
    "            target_line = target_line.split()\n",
    "            gene_ID = target_line[0]\n",
    "        #Many genes are not present, but could sway the averages if they are included later on, so they are removed in these cases\n",
    "        gap_removed = data.replace('-', '')\n",
    "        if len(gap_removed)>25:\n",
    "            average_length = average_length + len(gap_removed)\n",
    "        else:\n",
    "            missing=1\n",
    "        dyad_frequency = CpG/len(gap_removed)*100\n",
    "        #adds the data to the CSV        \n",
    "        file_num = 'Gene_'+str(line_num)\n",
    "\n",
    "        row_data = [file_num, gene_ID, species_name, len(gap_removed)-atg , CpG, dyad_frequency]    \n",
    "        writer.writerow(row_data)\n",
    "        #returns the data, most the data is re-entered for the next sequence, but must be returned like this to save it.  \n",
    "        return(dyad_dic, dyad_freq, dyad_species_name, gene_ID, dyad_ratio_dic, average_length, CpG, missing)\n",
    "\n",
    "#This is the second section (although technically it is where the code starts)\n",
    "#Generally speaking it is where the user input the data, and then the general document information is gathered and inputted into a csv\n",
    "#At the end of this section, you have the general document, and then a series of dictionaries that are then used to generate the individual locations document (section 3)\n",
    "\n",
    "    \n",
    "#This gets the list of gene names to Gene IDs\n",
    "with open('gene_correspondence.txt') as name_file:\n",
    "    content = name_file.readlines()\n",
    "directory = 'test_data'\n",
    "#empty object for later use\n",
    "gen_rows = {}\n",
    "gene_length=[]\n",
    "final_gene_ID=[]\n",
    "conserved_dic={}\n",
    "conserved_loc = {}\n",
    "length_dic ={}\n",
    "big_dic={}\n",
    "small_dic={}\n",
    "averages_dic={}\n",
    "animal_dic={}\n",
    "missing_dic={}\n",
    "# iterate over files in the test_data directory. Could be easily edited, or changed to be a user input\n",
    "for filename in os.listdir(directory):       \n",
    "#Creates the CSV with the desired headers  \n",
    "    with open('test_results/' + str(filename + '_general_results')+'.csv', 'w') as result:\n",
    "        headers = ['File Number', 'Gene ID', 'Species', 'Gene Length', 'No. of CpG', 'Dyad Freq. Factor']\n",
    "        writer = csv.writer(result)\n",
    "        writer.writerow(headers)\n",
    "        #empty objects needed for later on. \n",
    "        dyad_dic={}\n",
    "        dyad_ratio_dic={}\n",
    "        seq_list=[]\n",
    "        dic_info={}\n",
    "        temp_location_list={}\n",
    "        num_of_sequences = 0\n",
    "        dyad_freq={}\n",
    "        dyad_species_name={}       \n",
    "        file = filename\n",
    "        line_num=0\n",
    "        average_length =0 \n",
    "        #opens the file\n",
    "        with open('test_data/'+file) as f:\n",
    "            count=0\n",
    "            atg = atg_checker(file)\n",
    "            #For loop that goes over every line in the provided document. \n",
    "            for all_lines in f:\n",
    "                #If the line starts with a '>' then it is added as a new entry (as > indicates that it is a new sequence)\n",
    "                if all_lines[0] =='>':\n",
    "                    #This section trims the data of \\n, then runs the function which looks for dyads and then imports the data onto the first CSV. \n",
    "                    if count!=0:\n",
    "                        data = data.replace('\\n','').upper()\n",
    "                        fun_res = dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg)\n",
    "                        #These lines extract the data generated in the function above, the data will be needed for the section below. \n",
    "                        dyad_dic = fun_res[0]\n",
    "                        dyad_freq = fun_res[1]\n",
    "                        dyad_species_name = fun_res[2]\n",
    "                        temp_gene_ID = fun_res[3]\n",
    "                        dyad_ratio_dic = fun_res[4]\n",
    "                        average_length = fun_res[5]\n",
    "                        CpG=fun_res[6]\n",
    "                        missing=fun_res[7]\n",
    "                        num_of_sequences +=1\n",
    "                        if species_name in animal_dic.keys():\n",
    "                            animal_dic[species_name]=animal_dic[species_name]+CpG\n",
    "                            missing_dic[species_name]=missing_dic[species_name]+missing\n",
    "                        else:\n",
    "                            animal_dic[species_name]=CpG\n",
    "                            \n",
    "                            missing_dic[species_name]=missing\n",
    "                        \n",
    "                    #This is for the first line of the document, as there is no data to be run yet. \n",
    "                    elif count==0:\n",
    "                        count = 1\n",
    "                    #The next few lines get information about the gene that data is going to be collected for (specifically the Gene ID and species name)\n",
    "\n",
    "                    file_ID = str(file).replace('>','')\n",
    "                    species_name = str(all_lines.replace('>',''))\n",
    "                    species_name = str(species_name.replace('\\n',''))\n",
    "                    \n",
    "                    #Resets data to nothing for the next set of data and then continues onto the first line of data. \n",
    "                    data = ''\n",
    "                    continue\n",
    "                #Data is just data with the new line\n",
    "                data = data+all_lines\n",
    "            #As the end of the loop does not end on a >, the code for calculating dyads must be repeated once again \n",
    "            data = data.replace('\\n','').upper()\n",
    "            fun_res = dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg)\n",
    "            dyad_dic = fun_res[0]\n",
    "            dyad_freq = fun_res[1]\n",
    "            dyad_species_name = fun_res[2]                \n",
    "            num_of_sequences +=1\n",
    "            average_length = fun_res[5]\n",
    "            final_average = average_length/num_of_sequences\n",
    "            row_data = ['Average length of sequence', '', final_average, '', '']\n",
    "            writer.writerow(row_data)\n",
    "            averages_dic[temp_gene_ID] = final_average\n",
    "        \n",
    "        \n",
    "        \n",
    "        total_dyads=0\n",
    "        for each in dyad_freq.items():\n",
    "            total_dyads = each[1]+total_dyads\n",
    "        temp_final = {filename:total_dyads}\n",
    "        gen_rows.update(temp_final)\n",
    "        gene_length.append(len(data))\n",
    "        final_gene_ID.append(temp_gene_ID)\n",
    "\n",
    "#section 3        \n",
    "#If the inputted file has more then one sequence of DNA it will do the section below\n",
    "#This section looks at each location where there was at least 1 dyad and analyse it\n",
    "#It looks at the number dyads, the codon sequences, the amino acid sequence, the frequency of each codon, and then attempts to give an array of score that grade how likely the dyad pattern found is a result of chance. \n",
    "\n",
    "\n",
    "    #This section compared the dyads and looks for dyads that are highly conserved, and gives the species. It also looks at how likely that dyad sequence is conserved.\n",
    "    if num_of_sequences>1:\n",
    "            #Dictionary that can assign codons to amino acids\n",
    "        codon_to_amino = {\n",
    "    'Ala': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "    'Arg': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "    'Asn': ['AAT', 'AAC'],\n",
    "    'Asp': ['GAT', 'GAC'],\n",
    "    'Cys': ['TGT', 'TGC'],\n",
    "    'Glu': ['GAA', 'GAG'],\n",
    "    'Gln': ['CAA', 'CAG'],\n",
    "    'Gly': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "    'His': ['CAT', 'CAC'],\n",
    "    'Ile': ['ATT', 'ATC', 'ATA'],\n",
    "    'Leu': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "    'Lys': ['AAA', 'AAG'],\n",
    "    'Met': ['ATG'],\n",
    "    'Phe': ['TTT', 'TTC'],\n",
    "    'Pro': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "    'Ser': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "    'Thr': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "    'Trp': ['TGG'],\n",
    "    'Tyr': ['TAT', 'TAC'],\n",
    "    'Val': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "    'STOP': ['TAA', 'TAG', 'TGA']\n",
    "}\n",
    "\n",
    "            #dictionary with likelihood that each amino acid is conserved. \n",
    "        front_codon_dic = {'Phe': 24.71948300108424,\n",
    "                             'Leu': 6.231128355613783,\n",
    "                             'Ile': 27.856867589620553,\n",
    "                             'Met': 0.0,\n",
    "                             'Val': 7.107237039222716,\n",
    "                             'Ser': 21.207978547877214,\n",
    "                             'Pro': 17.20568849034703,\n",
    "                             'Thr': 14.975463866491085,\n",
    "                             'Ala': 20.088245011013637,\n",
    "                             'Tyr': 32.991588512980336,\n",
    "                             'His': 39.75633601626467,\n",
    "                             'Gln': 0.0,\n",
    "                             'Asn': 29.174469941162602,\n",
    "                             'Lys': 0.0,\n",
    "                             'Asp': 28.986902793614437,\n",
    "                             'Glu': 0.0,\n",
    "                             'Cys': 27.218103207070428,\n",
    "                             'Trp': 0.0,\n",
    "                             'Arg': 9.957816426216132,\n",
    "                             'Gly': 18.35457996927749}\n",
    "\n",
    "    #This writes a new CSV with the shared dyad data. \n",
    "        with open('test_results/' + str(file + '_dyad_results')+'.csv', 'w') as dyad_file:\n",
    "            #Sorts the core dictionary so it is index order\n",
    "            dyad_dic = dict(sorted(list(dyad_dic.items())))\n",
    "            dyad_ratio_dic = dict(sorted(dyad_ratio_dic.items()))\n",
    "            header = ['Gene ID','Dyad Index', 'Number of appearances', 'Dyad Sequences', 'Amino Acids', 'how frequent each amino acid appears','Ratio on how conserved the gene is.','Difference between expected and observed', 'Random Score', 'Mammal corrected score','Total Number of Sequences',  'Species present']\n",
    "            writer = csv.writer(dyad_file)\n",
    "            writer.writerow(header)\n",
    "            finaltemp=[]\n",
    "            templist=[]\n",
    "            count2=-1\n",
    "            conserved_dic[file]=0\n",
    "\n",
    "            for each in dyad_ratio_dic.items():\n",
    "                temp = list(each[1])\n",
    "                tempor = Counter(temp)\n",
    "                fulltemp=[]\n",
    "                for all in tempor.items():\n",
    "                    fulltemp.append(all[1])\n",
    "                finaltemp.append(fulltemp)\n",
    "\n",
    "\n",
    "            #For loop that goes over each core dictionary entry and extracts the relevant data. It also gives the shared index so that data cane be extracted from the other dictionaries.\n",
    "            for each_dyad in dyad_dic.items():\n",
    "                #redundancy_rate is set to 1 to start with, this is used later when observing how likely the codon pattern should be conserved. \n",
    "                redundancy_rate = 1\n",
    "                count2+=1\n",
    "                amino_list = []\n",
    "                list_of_redundancies=[]\n",
    "                #Extracting the information from dyad_dic to get the locations and the associated codon sequences\n",
    "                dyad_loc = each_dyad[0]\n",
    "                dyad_seqs = each_dyad[1]\n",
    "                #This looks into the dyad_freq dictionary and gets the final number of dyads at the associated location. \n",
    "                final_freq = dyad_freq[each_dyad[0]]\n",
    "                random_score = 0\n",
    "                #Gets the species name in the same way the final frequency is collected. \n",
    "                final_name = dyad_species_name[each_dyad[0]]\n",
    "                gene_code = final_gene_ID[-1]\n",
    "                if each_dyad[0] in conserved_loc.keys():\n",
    "\n",
    "                    conserved_loc[each_dyad[0]] = conserved_loc[each_dyad[0]] + final_freq\n",
    "                else:\n",
    "                    conserved_loc[each_dyad[0]] = final_freq\n",
    "                #This records if it is conserved dyad \n",
    "                if final_freq>200:\n",
    "                    conserved_dic[file]+=1\n",
    "                    fix_data = data.strip('-')\n",
    "                    temp_per = int((each_dyad[0]/len(fix_data))*100)\n",
    "                    if temp_per in length_dic:\n",
    "                        length_dic[temp_per]+=1\n",
    "                    else:\n",
    "                        length_dic[temp_per]=1\n",
    "                    #This checks if the gene fits into the large or small gene category\n",
    "                    if final_average>2000:\n",
    "                        if temp_per in big_dic:\n",
    "                            big_dic[temp_per]+=1\n",
    "                        else: \n",
    "                            big_dic[temp_per]=1\n",
    "                    else:\n",
    "                        if temp_per in small_dic:\n",
    "                            small_dic[temp_per]+=1\n",
    "                        else:\n",
    "                            small_dic[temp_per]=1\n",
    "                #The next for loop looks at all the codon pairs in dyad_seqs made above.\n",
    "                for codon_pair in dyad_seqs:\n",
    "                    #First the codon pair is split up to make the front and back codons\n",
    "                    front_codon = codon_pair[0:3]\n",
    "                    back_codon = codon_pair[3:6]\n",
    "                    #The next two for loops search for the codon sequences in the codon to amino dicitonary above and converts it to the relevant amino acid. \n",
    "                    for each_amino in codon_to_amino.items():\n",
    "                        #Some amino acids have multiple corisponding codons, in these cases a new for loop goes over each item and checks if front codon is in that list. \n",
    "                        if len(each_amino[1])>1:\n",
    "                            for list_amino in each_amino[1]:\n",
    "                                if front_codon == list_amino:               \n",
    "                                    #If the front codon is found in an amino's codon list then front amino is assigned to the corrisonding amino acid. \n",
    "                                    front_amino = each_amino[0]\n",
    "                        #If there is only one associated codon to an amino, then it just needs to do a basic check to see if it is present, then assign accordingly. \n",
    "                        if front_codon in each_amino:\n",
    "                            front_amino = each_amino[0]\n",
    "                    #This is a repeat of the above but looking at the back codon instead. \n",
    "                    for each_amino in codon_to_amino.items():\n",
    "                        if len(each_amino[1])>1:\n",
    "                            for list_amino in each_amino[1]:\n",
    "                                if back_codon == list_amino:                \n",
    "                                    back_amino = each_amino[0]\n",
    "                        if back_codon in each_amino:\n",
    "                            back_amino = each_amino[0]\n",
    "                    if [front_amino, back_amino] not in amino_list:\n",
    "                        amino_list.append([front_amino, back_amino])\n",
    "                    #A conserved ratio that indicates how conserved the gene is. \n",
    "                    conserved_ratio = final_freq/len(amino_list)\n",
    "                    for every in range(len(finaltemp[count2])):                            \n",
    "                        list_of_redundancies.append(front_codon_dic[front_amino])\n",
    "                #The final set of if statements calculate the likelihood that all the CpGs are conserved by chance. It also calculates a final score which is an arbitrar score but does indicate if a location has a hgih chance ot be interesting. \n",
    "                if final_freq== 1:\n",
    "                    redundancy_rate = 'NA'\n",
    "                    random_score = 0\n",
    "                    mammal_score=0 \n",
    "                else:\n",
    "                    redundancy_rate = float((sum(list_of_redundancies)/100)/len(list_of_redundancies))\n",
    "                    random_score = final_freq-(num_of_sequences*redundancy_rate)\n",
    "                    mammal_score = final_freq-(num_of_sequences*(float(redundancy_rate/10)))\n",
    "                codon_freq = list(dyad_ratio_dic.items())        \n",
    "                rows = [gene_code, dyad_loc, final_freq, dyad_seqs,amino_list, finaltemp[count2], conserved_ratio, redundancy_rate, random_score, mammal_score, num_of_sequences, final_name]\n",
    "                writer.writerow(rows)\n",
    "\n",
    "\n",
    "#Section 4 \n",
    "# The section generates the general documents \n",
    "# The first is a general document that has all genes, their respective number of CpGs, length of the gene, average number of CpGs per species, and a ratio value (number of CpGs/length of the gene)\n",
    "# The second document shows each dyad location, and how many conserved dyads are present (as in number of genes with a conserved dyad at that location)\n",
    "# The third document uses normalised data. Dyad location is transfomred into percentage along the gene length, frequency at each percentage is recorded, \n",
    "# and frequency is also recorded for small genes (<2000bp) and large genes (>2000) at each percentage point. \n",
    "\n",
    "#First is the general summary file                    \n",
    "counter=0\n",
    "#Initially, a file is generated\n",
    "with open(extension+'General_summary.csv', 'w') as gen:\n",
    "    headers = ['File name', 'Gene ID', 'Number of CpGs', 'Average number of CpGs', 'Length of Gene', 'Relative dyad concentration', 'How many highly conserved Ratio', 'Conservation against general presence']\n",
    "    writer = csv.writer(gen)\n",
    "    writer.writerow(headers)\n",
    "    \n",
    "    #For loop for each gene, with associated number of dyads. \n",
    "    for each_gene in gen_rows.items():\n",
    "        # ratio = relative dyad concentration\n",
    "        ratio = each_gene[1]/gene_length[counter]\n",
    "        num_conserved_ratio = (conserved_dic[each_gene[0]]/gene_length[counter])*100\n",
    "        average_gene_length = averages_dic[final_gene_ID[counter]]\n",
    "        if num_conserved_ratio == 0:\n",
    "            ratio_ratio=0\n",
    "        else:\n",
    "            #Ratio_ratio = relative dyad concentration divided by the relative number of conserved dyads\n",
    "            # essnetially, there may be many dyads or not many conserved, or there are few dyads but they are conserved. \n",
    "            ratio_ratio = ratio/num_conserved_ratio\n",
    "        #This is a list of data that forms one line, which is then written into the document made. \n",
    "        fin = [each_gene[0], final_gene_ID[counter], each_gene[1], (each_gene[1]/261), average_gene_length, ratio, num_conserved_ratio, ratio_ratio]\n",
    "        writer.writerow(fin)\n",
    "        counter+=1\n",
    "\n",
    "#The document detailed locations of conserved dyads is generated\n",
    "with open(extension+'conserved_locations.csv', 'w') as conserve:\n",
    "    headers = ['location', 'no.of conserved dyads']\n",
    "    writer = csv.writer(conserve)\n",
    "    writer.writerow(headers)\n",
    "    #The conserved locations dictionaries first needs to be sorted into numerical order\n",
    "    conserved_loc = dict(sorted(conserved_loc.items()))\n",
    "    #for loop going over each conserved location and writing a line into the document show the index (ind) and the total number of conserved dyads (tot)\n",
    "    for each_conserved_loc in conserved_loc.items():\n",
    "        ind = each_conserved_loc[0]\n",
    "        tot = each_conserved_loc[1]\n",
    "        fin = [ind, tot]\n",
    "        writer.writerow(fin)\n",
    "\n",
    "#Final document has data on the positions of conserved dyads along a normalised percentage gradient (percentage long the gene length)\n",
    "#It also looks if it is a large or small gene and assigns it accordingly (allowing comparisons between the two sizes) \n",
    "with open(extension+'percentage_locations.csv', 'w') as per_con:\n",
    "    headers = ['Percent', 'Freq', 'big', 'small']\n",
    "    writer = csv.writer(per_con)\n",
    "    writer.writerow(headers)\n",
    "    #Dictionaries are sorted into percentage order, from 0-99\n",
    "    length_dic = dict(sorted(length_dic.items()))\n",
    "    big_dic = dict(sorted(big_dic.items()))\n",
    "    small_dic = dict(sorted(small_dic.items()))\n",
    "    #for loop that looks at each percentage point and records the number of dyads at each point\n",
    "    for each_length_percentage in length_dic.items():\n",
    "        ind = each_length_percentage[0]\n",
    "        tot = each_length_percentage[1]\n",
    "        #Errors can be thrown if there are 0 dyads for a certain size, so an If statement is needed to catch said errors\n",
    "        if each_length_percentage[0] not in big_dic.keys():\n",
    "            big_val=0\n",
    "        else:\n",
    "            big_val = big_dic[each_length_percentage[0]]\n",
    "        if each_length_percentage[0] not in small_dic.keys():\n",
    "            small_val=0\n",
    "        else:\n",
    "            small_val = small_dic[each_length_percentage[0]]\n",
    "        fin = [ind, tot, big_val, small_val]\n",
    "        writer.writerow(fin)\n",
    "        \n",
    "with open(extension+'animal__frequency.csv', 'w') as ani_freq:\n",
    "    headers = ['species', 'Number of CpGs', 'Number of missing genes']\n",
    "    writer= csv.writer(ani_freq)\n",
    "    writer.writerow(headers)\n",
    "    for each in animal_dic.items():\n",
    "        animal = each[0]\n",
    "        freq = each[1]\n",
    "        miss = missing_dic[each[0]]\n",
    "        fin = [animal, freq, miss]\n",
    "        writer.writerow(fin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02bcd10d-23f2-4f7b-b415-b274d2c7aebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "###This version asks for an inputted gene list. It is otherwise the same as the above. ###\n",
    "###################################\n",
    "\n",
    "# This script will analyse all sequences in the 'test data' folder.\n",
    "# Currently it is only set up for this specific data set, meaning a FASTA file named in the way the example data is\n",
    "# final results for individual genes are outputted into the test_results folder, a few general result documents are also generated, the are outputted to the source folder.\n",
    "# The code could be adapted to look for .txt files or other files with some effort. \n",
    "\n",
    "### Instructions!\n",
    "\n",
    "# This is very simple, put all test data into the test_data folder, then hit ctrl+enter\n",
    "\n",
    "# The only thing that is worth mentioning is that to match the ENSG gene code to the gene number there is a gene_correspondents.txt file,\n",
    "# If you add new sequences that do not have a gene ID then you will likely get an error. I can edit the code so that it gives an 'NA' rather then an error if needed. \n",
    "\n",
    "#You can enter an extension name for the final results files if you like, otherwise it can be left blank\n",
    "extension='NNCGNN_'\n",
    "\n",
    "###For the input, it is tab seperated. I usually achieve this in excel by having a file for each line.\n",
    "### Also note that I input file names, not gene or ENGS names\n",
    "\n",
    "\n",
    "# This code is split into 4 sections, the first three are the same in all three codes. \n",
    "\n",
    "\n",
    "# First is the function used for identifying dyads\n",
    "# I recommend against changing this unless you want the general information document to present something different\n",
    "# If you do add or change data, you will need to extract it in the second section\n",
    "\n",
    "#Needed for generating the csv\n",
    "import csv\n",
    "#os is needed to iterate over the folder so you can do all data points there in \n",
    "import os\n",
    "#A counting package needed further on. \n",
    "from collections import Counter\n",
    "\n",
    "#Many genes have a double ATG at the start (nessecary in the previous study, but unwanted here)\n",
    "#This function checks if the gene has the double ATG, and records it. This allows for correction later on.\n",
    "def atg_checker(file):\n",
    "    with open('test_data/'+file) as f2:\n",
    "        lines=f2.read()\n",
    "        lines=lines.split('\\n')\n",
    "        count2=0\n",
    "        for line in lines:\n",
    "            if len(line)==0 or line[0]=='>':\n",
    "                continue\n",
    "            elif line[0:6]=='ATGATG' or line[0:6] =='ATG---':\n",
    "                count2+=1\n",
    "            else:\n",
    "                continue\n",
    "        if count2 ==261:\n",
    "            return(3)\n",
    "        else:\n",
    "            return(0)\n",
    "\n",
    "\n",
    "#Function for counting Dyads, more functionality can easily be added if needed. \n",
    "def dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg):\n",
    "        #Create/reset objects\n",
    "        i = 0\n",
    "        CpG = 0\n",
    "        dyad_locations = []\n",
    "        dyad_seq=[]\n",
    "        line_num=''\n",
    "        gene_ID=''\n",
    "        missing=0\n",
    "        #A while loop that goes up in codons and checks between the codons for CpGs, it then counts these.\n",
    "        while i< len(data):\n",
    "            #counts up the index each loop\n",
    "            i=i+3\n",
    "            #checks if the index is above the size of the sequence after the addition. \n",
    "            if i>=len(data):\n",
    "                break\n",
    "        #checks if both nessecary bases are C and G, if so it adds 1 to the count. \n",
    "            if (data[i]=='G' and data[i-1]=='C'): #or (data[i]=='G' and data[i-1]=='G' and data[i-2]=='C'): #or (data[i-1]=='G' and data[i-2]=='C') #edit depending on what you want to analyse.\n",
    "                CpG+=1\n",
    "                codons = data[i-3:i+3]\n",
    "                \n",
    "                #This section updates the 3 dictionaries with the data we need for analysing the codons. \n",
    "                #If the dyad position is already in the main dicitonary then it needs to update the relevant dictionaries\n",
    "                if i-atg in dyad_dic:\n",
    "                    #These lines add the species name to the species name dictionary, this data is mainly visual for the end\n",
    "                    temp_name_list = dyad_species_name[i-atg]\n",
    "                    temp_name_list.append(species_name)\n",
    "                    temp_name_dic = {i-atg:temp_name_list}\n",
    "                    dyad_species_name.update(temp_name_dic)\n",
    "                    #The next dictionary check is to see if the location's sequence is already present, if codons are conserved then is will happen a lot, it adds to the frequency then moves to the next codon\n",
    "                    if codons in dyad_dic[i-atg]:\n",
    "                        dyad_freq[i-atg]+=1\n",
    "                        second_temp=dyad_ratio_dic[i-atg]                                \n",
    "                        second_temp.append(codons)\n",
    "                        #print(second_temp)\n",
    "                        temp_dic = {i-atg:second_temp}\n",
    "                        dyad_ratio_dic.update(temp_dic)\n",
    "                        continue\n",
    "                    #If the codon is not present, then a new codon is added to entry at the given position\n",
    "                    else:\n",
    "                        temp_list=[]\n",
    "                        second_temp=[]\n",
    "                        for num in dyad_dic.items():\n",
    "                            if num[0]==i-atg:\n",
    "                                dyad_freq[i-atg]+=1\n",
    "                                temp_list=dyad_dic[i-atg]\n",
    "                                temp_list.append(codons)\n",
    "                                temp_dic = {i-atg:temp_list}\n",
    "                                dyad_dic.update(temp_dic)\n",
    "                                second_temp=dyad_ratio_dic[i-atg]                                \n",
    "                                second_temp.append(codons)\n",
    "                                #print(second_temp)\n",
    "                                temp_dic = {i-atg:second_temp}\n",
    "                                dyad_ratio_dic.update(temp_dic)\n",
    "                                \n",
    "                #If the entry is new, then the first set of data for the three dictionaries is entered.                \n",
    "                else:\n",
    "                    dyad_dic[i-atg]=[codons]\n",
    "                    dyad_freq[i-atg]=1\n",
    "                    dyad_species_name[i-atg] = [species_name]\n",
    "                    dyad_ratio_dic[i-atg]=[codons]\n",
    "                    \n",
    "        #This code turns the file_ID into a number, it then uses that number to get teh corresponding line in gene_correspondents, which then gives it the gene code which is recorded. \n",
    "        if file_ID !='':\n",
    "            line_num = file_ID.strip('.fasta')\n",
    "            line_num = int(line_num.strip('gene_'))\n",
    "\n",
    "            target_line = content[line_num-1]\n",
    "            target_line = target_line.split()\n",
    "            gene_ID = target_line[0]\n",
    "        #Many genes are not present, but could sway the averages if they are included later on, so they are removed in these cases\n",
    "        gap_removed = data.replace('-', '')\n",
    "        if len(gap_removed)>25:\n",
    "            average_length = average_length + len(gap_removed)\n",
    "        else:\n",
    "            missing=1\n",
    "        dyad_frequency = CpG/len(gap_removed)*100\n",
    "        #adds the data to the CSV        \n",
    "        file_num = 'Gene_'+str(line_num)\n",
    "\n",
    "        row_data = [file_num, gene_ID, species_name, len(gap_removed)-atg , CpG, dyad_frequency]    \n",
    "        writer.writerow(row_data)\n",
    "        #returns the data, most the data is re-entered for the next sequence, but must be returned like this to save it.  \n",
    "        return(dyad_dic, dyad_freq, dyad_species_name, gene_ID, dyad_ratio_dic, average_length, CpG, missing)\n",
    "\n",
    "#This is the second section (although technically it is where the code starts)\n",
    "#Generally speaking it is where the user input the data, and then the general document information is gathered and inputted into a csv\n",
    "#At the end of this section, you have the general document, and then a series of dictionaries that are then used to generate the individual locations document (section 3)\n",
    "\n",
    "    \n",
    "#This gets the list of gene names to Gene IDs\n",
    "with open('gene_correspondence.txt') as name_file:\n",
    "    content = name_file.readlines()\n",
    "directory = 'test_data'\n",
    "#This is the line that allows the input. If you have your list seperated in a different way then it is this that needs to change.\n",
    "list_input = list(input().split(' '))\n",
    "#empty object for later use\n",
    "gen_rows = {}\n",
    "gene_length=[]\n",
    "final_gene_ID=[]\n",
    "conserved_dic={}\n",
    "conserved_loc = {}\n",
    "length_dic ={}\n",
    "big_dic={}\n",
    "small_dic={}\n",
    "averages_dic={}\n",
    "animal_dic={}\n",
    "missing_dic={}\n",
    "# iterate over files in the test_data directory. Could be easily edited, or changed to be a user input\n",
    "for filename in os.listdir(directory):\n",
    "    #Quick if statement to check if the file is in the specifies list. \n",
    "    if filename not in list_input:\n",
    "        continue\n",
    "#Creates the CSV with the desired headers  \n",
    "    with open('test_results/' + str(filename + '_general_results')+'.csv', 'w') as result:\n",
    "        headers = ['File Number', 'Gene ID', 'Species', 'Gene Length', 'No. of CpG', 'Dyad Freq. Factor']\n",
    "        writer = csv.writer(result)\n",
    "        writer.writerow(headers)\n",
    "        #empty objects needed for later on. \n",
    "        dyad_dic={}\n",
    "        dyad_ratio_dic={}\n",
    "        seq_list=[]\n",
    "        dic_info={}\n",
    "        temp_location_list={}\n",
    "        num_of_sequences = 0\n",
    "        dyad_freq={}\n",
    "        dyad_species_name={}       \n",
    "        file = filename\n",
    "        line_num=0\n",
    "        average_length =0 \n",
    "        #opens the file\n",
    "        with open('test_data/'+file) as f:\n",
    "            count=0\n",
    "            atg = atg_checker(file)\n",
    "            #For loop that goes over every line in the provided document. \n",
    "            for all_lines in f:\n",
    "                #If the line starts with a '>' then it is added as a new entry (as > indicates that it is a new sequence)\n",
    "                if all_lines[0] =='>':\n",
    "                    #This section trims the data of \\n, then runs the function which looks for dyads and then imports the data onto the first CSV. \n",
    "                    if count!=0:\n",
    "                        data = data.replace('\\n','').upper()\n",
    "                        fun_res = dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg)\n",
    "                        #These lines extract the data generated in the function above, the data will be needed for the section below. \n",
    "                        dyad_dic = fun_res[0]\n",
    "                        dyad_freq = fun_res[1]\n",
    "                        dyad_species_name = fun_res[2]\n",
    "                        temp_gene_ID = fun_res[3]\n",
    "                        dyad_ratio_dic = fun_res[4]\n",
    "                        average_length = fun_res[5]\n",
    "                        CpG=fun_res[6]\n",
    "                        missing=fun_res[7]\n",
    "                        num_of_sequences +=1\n",
    "                        if species_name in animal_dic.keys():\n",
    "                            animal_dic[species_name]=animal_dic[species_name]+CpG\n",
    "                            missing_dic[species_name]=missing_dic[species_name]+missing\n",
    "                        else:\n",
    "                            animal_dic[species_name]=CpG\n",
    "                            \n",
    "                            missing_dic[species_name]=missing\n",
    "                        \n",
    "                    #This is for the first line of the document, as there is no data to be run yet. \n",
    "                    elif count==0:\n",
    "                        count = 1\n",
    "                    #The next few lines get information about the gene that data is going to be collected for (specifically the Gene ID and species name)\n",
    "\n",
    "                    file_ID = str(file).replace('>','')\n",
    "                    species_name = str(all_lines.replace('>',''))\n",
    "                    species_name = str(species_name.replace('\\n',''))\n",
    "                    \n",
    "                    #Resets data to nothing for the next set of data and then continues onto the first line of data. \n",
    "                    data = ''\n",
    "                    continue\n",
    "                #Data is just data with the new line\n",
    "                data = data+all_lines\n",
    "            #As the end of the loop does not end on a >, the code for calculating dyads must be repeated once again \n",
    "            data = data.replace('\\n','').upper()\n",
    "            fun_res = dyad_counter(data, file_ID, species_name, dyad_dic, dyad_freq, dyad_species_name, dyad_ratio_dic, average_length,atg)\n",
    "            dyad_dic = fun_res[0]\n",
    "            dyad_freq = fun_res[1]\n",
    "            dyad_species_name = fun_res[2]                \n",
    "            num_of_sequences +=1\n",
    "            average_length = fun_res[5]\n",
    "            final_average = average_length/num_of_sequences\n",
    "            row_data = ['Average length of sequence', '', final_average, '', '']\n",
    "            writer.writerow(row_data)\n",
    "            averages_dic[temp_gene_ID] = final_average\n",
    "        \n",
    "        \n",
    "        \n",
    "        total_dyads=0\n",
    "        for each in dyad_freq.items():\n",
    "            total_dyads = each[1]+total_dyads\n",
    "        temp_final = {filename:total_dyads}\n",
    "        gen_rows.update(temp_final)\n",
    "        gene_length.append(len(data))\n",
    "        final_gene_ID.append(temp_gene_ID)\n",
    "\n",
    "#section 3        \n",
    "#If the inputted file has more then one sequence of DNA it will do the section below\n",
    "#This section looks at each location where there was at least 1 dyad and analyse it\n",
    "#It looks at the number dyads, the codon sequences, the amino acid sequence, the frequency of each codon, and then attempts to give an array of score that grade how likely the dyad pattern found is a result of chance. \n",
    "\n",
    "\n",
    "    #This section compared the dyads and looks for dyads that are highly conserved, and gives the species. It also looks at how likely that dyad sequence is conserved.\n",
    "    if num_of_sequences>1:\n",
    "            #Dictionary that can assign codons to amino acids\n",
    "        codon_to_amino = {\n",
    "    'Ala': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "    'Arg': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "    'Asn': ['AAT', 'AAC'],\n",
    "    'Asp': ['GAT', 'GAC'],\n",
    "    'Cys': ['TGT', 'TGC'],\n",
    "    'Glu': ['GAA', 'GAG'],\n",
    "    'Gln': ['CAA', 'CAG'],\n",
    "    'Gly': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "    'His': ['CAT', 'CAC'],\n",
    "    'Ile': ['ATT', 'ATC', 'ATA'],\n",
    "    'Leu': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "    'Lys': ['AAA', 'AAG'],\n",
    "    'Met': ['ATG'],\n",
    "    'Phe': ['TTT', 'TTC'],\n",
    "    'Pro': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "    'Ser': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "    'Thr': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "    'Trp': ['TGG'],\n",
    "    'Tyr': ['TAT', 'TAC'],\n",
    "    'Val': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "    'STOP': ['TAA', 'TAG', 'TGA']\n",
    "}\n",
    "\n",
    "            #dictionary with likelihood that each amino acid is conserved. \n",
    "        front_codon_dic = {'Phe': 24.71948300108424,\n",
    "                             'Leu': 6.231128355613783,\n",
    "                             'Ile': 27.856867589620553,\n",
    "                             'Met': 0.0,\n",
    "                             'Val': 7.107237039222716,\n",
    "                             'Ser': 21.207978547877214,\n",
    "                             'Pro': 17.20568849034703,\n",
    "                             'Thr': 14.975463866491085,\n",
    "                             'Ala': 20.088245011013637,\n",
    "                             'Tyr': 32.991588512980336,\n",
    "                             'His': 39.75633601626467,\n",
    "                             'Gln': 0.0,\n",
    "                             'Asn': 29.174469941162602,\n",
    "                             'Lys': 0.0,\n",
    "                             'Asp': 28.986902793614437,\n",
    "                             'Glu': 0.0,\n",
    "                             'Cys': 27.218103207070428,\n",
    "                             'Trp': 0.0,\n",
    "                             'Arg': 9.957816426216132,\n",
    "                             'Gly': 18.35457996927749}\n",
    "\n",
    "    #This writes a new CSV with the shared dyad data. \n",
    "        with open('test_results/' + str(file + '_dyad_results')+'.csv', 'w') as dyad_file:\n",
    "            #Sorts the core dictionary so it is index order\n",
    "            dyad_dic = dict(sorted(list(dyad_dic.items())))\n",
    "            dyad_ratio_dic = dict(sorted(dyad_ratio_dic.items()))\n",
    "            header = ['Gene ID','Dyad Index', 'Number of appearances', 'Dyad Sequences', 'Amino Acids', 'how frequent each amino acid appears','Ratio on how conserved the gene is.','Difference between expected and observed', 'Random Score', 'Mammal corrected score','Total Number of Sequences',  'Species present']\n",
    "            writer = csv.writer(dyad_file)\n",
    "            writer.writerow(header)\n",
    "            finaltemp=[]\n",
    "            templist=[]\n",
    "            count2=-1\n",
    "            conserved_dic[file]=0\n",
    "\n",
    "            for each in dyad_ratio_dic.items():\n",
    "                temp = list(each[1])\n",
    "                tempor = Counter(temp)\n",
    "                fulltemp=[]\n",
    "                for all in tempor.items():\n",
    "                    fulltemp.append(all[1])\n",
    "                finaltemp.append(fulltemp)\n",
    "\n",
    "\n",
    "            #For loop that goes over each core dictionary entry and extracts the relevant data. It also gives the shared index so that data cane be extracted from the other dictionaries.\n",
    "            for each_dyad in dyad_dic.items():\n",
    "                #redundancy_rate is set to 1 to start with, this is used later when observing how likely the codon pattern should be conserved. \n",
    "                redundancy_rate = 1\n",
    "                count2+=1\n",
    "                amino_list = []\n",
    "                list_of_redundancies=[]\n",
    "                #Extracting the information from dyad_dic to get the locations and the associated codon sequences\n",
    "                dyad_loc = each_dyad[0]\n",
    "                dyad_seqs = each_dyad[1]\n",
    "                #This looks into the dyad_freq dictionary and gets the final number of dyads at the associated location. \n",
    "                final_freq = dyad_freq[each_dyad[0]]\n",
    "                random_score = 0\n",
    "                #Gets the species name in the same way the final frequency is collected. \n",
    "                final_name = dyad_species_name[each_dyad[0]]\n",
    "                gene_code = final_gene_ID[-1]\n",
    "                if each_dyad[0] in conserved_loc.keys():\n",
    "\n",
    "                    conserved_loc[each_dyad[0]] = conserved_loc[each_dyad[0]] + final_freq\n",
    "                else:\n",
    "                    conserved_loc[each_dyad[0]] = final_freq\n",
    "                #This records if it is conserved dyad \n",
    "                if final_freq>200:\n",
    "                    conserved_dic[file]+=1\n",
    "                    fix_data = data.strip('-')\n",
    "                    temp_per = int((each_dyad[0]/len(fix_data))*100)\n",
    "                    if temp_per in length_dic:\n",
    "                        length_dic[temp_per]+=1\n",
    "                    else:\n",
    "                        length_dic[temp_per]=1\n",
    "                    #This checks if the gene fits into the large or small gene category\n",
    "                    if final_average>2000:\n",
    "                        if temp_per in big_dic:\n",
    "                            big_dic[temp_per]+=1\n",
    "                        else: \n",
    "                            big_dic[temp_per]=1\n",
    "                    else:\n",
    "                        if temp_per in small_dic:\n",
    "                            small_dic[temp_per]+=1\n",
    "                        else:\n",
    "                            small_dic[temp_per]=1\n",
    "                #The next for loop looks at all the codon pairs in dyad_seqs made above.\n",
    "                for codon_pair in dyad_seqs:\n",
    "                    #First the codon pair is split up to make the front and back codons\n",
    "                    front_codon = codon_pair[0:3]\n",
    "                    back_codon = codon_pair[3:6]\n",
    "                    #The next two for loops search for the codon sequences in the codon to amino dicitonary above and converts it to the relevant amino acid. \n",
    "                    for each_amino in codon_to_amino.items():\n",
    "                        #Some amino acids have multiple corisponding codons, in these cases a new for loop goes over each item and checks if front codon is in that list. \n",
    "                        if len(each_amino[1])>1:\n",
    "                            for list_amino in each_amino[1]:\n",
    "                                if front_codon == list_amino:               \n",
    "                                    #If the front codon is found in an amino's codon list then front amino is assigned to the corrisonding amino acid. \n",
    "                                    front_amino = each_amino[0]\n",
    "                        #If there is only one associated codon to an amino, then it just needs to do a basic check to see if it is present, then assign accordingly. \n",
    "                        if front_codon in each_amino:\n",
    "                            front_amino = each_amino[0]\n",
    "                    #This is a repeat of the above but looking at the back codon instead. \n",
    "                    for each_amino in codon_to_amino.items():\n",
    "                        if len(each_amino[1])>1:\n",
    "                            for list_amino in each_amino[1]:\n",
    "                                if back_codon == list_amino:                \n",
    "                                    back_amino = each_amino[0]\n",
    "                        if back_codon in each_amino:\n",
    "                            back_amino = each_amino[0]\n",
    "                    if [front_amino, back_amino] not in amino_list:\n",
    "                        amino_list.append([front_amino, back_amino])\n",
    "                    #A conserved ratio that indicates how conserved the gene is. \n",
    "                    conserved_ratio = final_freq/len(amino_list)\n",
    "                    for every in range(len(finaltemp[count2])):                            \n",
    "                        list_of_redundancies.append(front_codon_dic[front_amino])\n",
    "                #The final set of if statements calculate the likelihood that all the CpGs are conserved by chance. It also calculates a final score which is an arbitrar score but does indicate if a location has a hgih chance ot be interesting. \n",
    "                if final_freq== 1:\n",
    "                    redundancy_rate = 'NA'\n",
    "                    random_score = 0\n",
    "                    mammal_score=0 \n",
    "                else:\n",
    "                    redundancy_rate = float((sum(list_of_redundancies)/100)/len(list_of_redundancies))\n",
    "                    random_score = final_freq-(num_of_sequences*redundancy_rate)\n",
    "                    mammal_score = final_freq-(num_of_sequences*(float(redundancy_rate/10)))\n",
    "                codon_freq = list(dyad_ratio_dic.items())        \n",
    "                rows = [gene_code, dyad_loc, final_freq, dyad_seqs,amino_list, finaltemp[count2], conserved_ratio, redundancy_rate, random_score, mammal_score, num_of_sequences, final_name]\n",
    "                writer.writerow(rows)\n",
    "\n",
    "\n",
    "#Section 4 \n",
    "# The section generates the general documents \n",
    "# The first is a general document that has all genes, their respective number of CpGs, length of the gene, average number of CpGs per species, and a ratio value (number of CpGs/length of the gene)\n",
    "# The second document shows each dyad location, and how many conserved dyads are present (as in number of genes with a conserved dyad at that location)\n",
    "# The third document uses normalised data. Dyad location is transfomred into percentage along the gene length, frequency at each percentage is recorded, \n",
    "# and frequency is also recorded for small genes (<2000bp) and large genes (>2000) at each percentage point. \n",
    "\n",
    "#First is the general summary file                    \n",
    "counter=0\n",
    "#Initially, a file is generated\n",
    "with open(extension+'General_summary.csv', 'w') as gen:\n",
    "    headers = ['File name', 'Gene ID', 'Number of CpGs', 'Average number of CpGs', 'Length of Gene', 'Relative dyad concentration', 'How many highly conserved Ratio', 'Conservation against general presence']\n",
    "    writer = csv.writer(gen)\n",
    "    writer.writerow(headers)\n",
    "    \n",
    "    #For loop for each gene, with associated number of dyads. \n",
    "    for each_gene in gen_rows.items():\n",
    "        # ratio = relative dyad concentration\n",
    "        ratio = each_gene[1]/gene_length[counter]\n",
    "        num_conserved_ratio = (conserved_dic[each_gene[0]]/gene_length[counter])*100\n",
    "        average_gene_length = averages_dic[final_gene_ID[counter]]\n",
    "        if num_conserved_ratio == 0:\n",
    "            ratio_ratio=0\n",
    "        else:\n",
    "            #Ratio_ratio = relative dyad concentration divided by the relative number of conserved dyads\n",
    "            # essnetially, there may be many dyads or not many conserved, or there are few dyads but they are conserved. \n",
    "            ratio_ratio = ratio/num_conserved_ratio\n",
    "        #This is a list of data that forms one line, which is then written into the document made. \n",
    "        fin = [each_gene[0], final_gene_ID[counter], each_gene[1], (each_gene[1]/261), average_gene_length, ratio, num_conserved_ratio, ratio_ratio]\n",
    "        writer.writerow(fin)\n",
    "        counter+=1\n",
    "\n",
    "#The document detailed locations of conserved dyads is generated\n",
    "with open(extension+'conserved_locations.csv', 'w') as conserve:\n",
    "    headers = ['location', 'no.of conserved dyads']\n",
    "    writer = csv.writer(conserve)\n",
    "    writer.writerow(headers)\n",
    "    #The conserved locations dictionaries first needs to be sorted into numerical order\n",
    "    conserved_loc = dict(sorted(conserved_loc.items()))\n",
    "    #for loop going over each conserved location and writing a line into the document show the index (ind) and the total number of conserved dyads (tot)\n",
    "    for each_conserved_loc in conserved_loc.items():\n",
    "        ind = each_conserved_loc[0]\n",
    "        tot = each_conserved_loc[1]\n",
    "        fin = [ind, tot]\n",
    "        writer.writerow(fin)\n",
    "\n",
    "#Final document has data on the positions of conserved dyads along a normalised percentage gradient (percentage long the gene length)\n",
    "#It also looks if it is a large or small gene and assigns it accordingly (allowing comparisons between the two sizes) \n",
    "with open(extension+'percentage_locations.csv', 'w') as per_con:\n",
    "    headers = ['Percent', 'Freq', 'big', 'small']\n",
    "    writer = csv.writer(per_con)\n",
    "    writer.writerow(headers)\n",
    "    #Dictionaries are sorted into percentage order, from 0-99\n",
    "    length_dic = dict(sorted(length_dic.items()))\n",
    "    big_dic = dict(sorted(big_dic.items()))\n",
    "    small_dic = dict(sorted(small_dic.items()))\n",
    "    #for loop that looks at each percentage point and records the number of dyads at each point\n",
    "    for each_length_percentage in length_dic.items():\n",
    "        ind = each_length_percentage[0]\n",
    "        tot = each_length_percentage[1]\n",
    "        #Errors can be thrown if there are 0 dyads for a certain size, so an If statement is needed to catch said errors\n",
    "        if each_length_percentage[0] not in big_dic.keys():\n",
    "            big_val=0\n",
    "        else:\n",
    "            big_val = big_dic[each_length_percentage[0]]\n",
    "        if each_length_percentage[0] not in small_dic.keys():\n",
    "            small_val=0\n",
    "        else:\n",
    "            small_val = small_dic[each_length_percentage[0]]\n",
    "        fin = [ind, tot, big_val, small_val]\n",
    "        writer.writerow(fin)\n",
    "        \n",
    "with open(extension+'animal__frequency.csv', 'w') as ani_freq:\n",
    "    headers = ['species', 'Number of CpGs', 'Number of missing genes']\n",
    "    writer= csv.writer(ani_freq)\n",
    "    writer.writerow(headers)\n",
    "    for each in animal_dic.items():\n",
    "        animal = each[0]\n",
    "        freq = each[1]\n",
    "        miss = missing_dic[each[0]]\n",
    "        fin = [animal, freq, miss]\n",
    "        writer.writerow(fin)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
